### ISSUE TITLE
@interfaceObject: Implementing type of an Interface Entity now having access to field extended on the Interface Entity

### ISSUE THREAD

**Issue by moonflare** [apollographql/federation#2548]
### Issue Description

When we define an Entity Interface with some implementing types, extend the Interface Entity in a subgraph with more fields, but reference just one Implementing Type Entity (Not the Interface Entity) from another subgraph, that entity type can't resolve the extra fields extended to the Interface Entity. For example, if we take the schema below:

```GraphQL
  ## Product Subgraph
  
  interface Product @key(fields: "id") {
    id: ID!
    name: String!
    price: Float
  }

  type Book implements Product @key(fields: "id") {
    id: ID!
    name: String!
    price: Float
    pages: Int
  }

  type Movie implements Product @key(fields: "id") {
    id: ID!
    name: String!
    price: Float
    duration: Int
  }

  ## Inventory Subgraph
  
  type Product @key(fields: "id") @interfaceObject {
    id: ID! @external
    price: Float @external
    inStock: Boolean
    shippingEstimate: Int @requires(fields: "price")
  }
 
  ## Reviews subgraph
  type Review @key(fields: "id") {
    id: ID!
    body: String
    product: Book
  }

  type Book @key(fields: "id") {
    id: ID!
    reviews: [Review]
  }
```

We will get the following error:
```bash
Error: A valid schema couldn't be composed. The following composition errors were found:
        The following supergraph API query:
{
  topProducts(first: 0) {
    ... on Movie {
      shippingEstimate
    }
  }
}
cannot be satisfied by the subgraphs because:
- from subgraph "products": cannot find field "Movie.shippingEstimate".
- from subgraph "inventory": cannot satisfy @require conditions on field "Product.shippingEstimate" (please ensure that this is not due to key field "id" being accidentally marked @external).
        The following supergraph API query:
{
  topProducts(first: 0) {
    ... on Book {
      shippingEstimate
    }
  }
}
cannot be satisfied by the subgraphs because:
- from subgraph "products": cannot find field "Book.shippingEstimate".
- from subgraph "inventory": cannot satisfy @require conditions on field "Product.shippingEstimate" (please ensure that this is not due to key field "id" being accidentally marked @external).
- from subgraph "reviews":
  - cannot find field "Book.shippingEstimate".
  - ignoring moving to subgraph "inventory" using @key(fields: "id") of "Book" because there is a more direct path in inventory that avoids reviews altogether.
```

### Expectations
I would expect the `Book` Entity referenced in the `Reviews subgraph` to be aware of the fields added to the `Product` Entity Interface in the Inventory subgraph.

### Link to Reproduction

https://github.com/moonflare/federation-demo/pull/1/commits/d47120d88874060704e78935f55df4c5cd9ae92a

### Reproduction Steps

```bash
## Install npm delepndencies
npm i

## Start the services
npm run start-services

## Start the Gateway
npm run start-gateway
```

**Comment by jeffjakub** [comment]
Hi Dan, Thanks for raising this issue and the detailed steps for how to reproduce the issue! Our team will look into it and get back to you as soon as possible.

**Comment by pcmanus** [comment]
The problem here is that `Product.id` field in the `Inventory` subgraph should _not_ be marked `@external`.

I'll explain, but I'll first note that the underlying problem is not specific to `@interfaceObject` at all, and you would get the exact same problem without it.

I think you ran into this due to some unfortunate subtlety that exists because of backward compatibility. Namely, looking at [the commit you link for the reproduction](https://github.com/moonflare/federation-demo/commit/d47120d88874060704e78935f55df4c5cd9ae92a), it seems that `Product` in `Inventory` used to be defined with an `extend type`, but you decided to remove that `extend`. But as you do so, you _also_ need to remove the `@external` on the key fields simultaneously. The reason (and subtlety here) is that `@external` _on key fields_ of types defined with `extend` is *ignored* (because that is what federation 1 was doing, and is why this is backward compatibility thing), but importantly, it is _not_ ignored if the type has no `extend`. So by removing the `extend` on the type but not the `@external` on the key field, you essentially _added_ `@external` on the key field. And while putting `@external` on key field can have uses (very corner case though), it's not an option here due to the `@requires`.

I understand this is a bit confusing, but to rephrase this maybe a bit more simply:
- in federation 1, you had to use `extend type` for all but the "owning" subgraph, and you _had_ to use `@external` on the key fields of those `extend type` _even though_ those key field had to be resolved in general. So while you had to mark them `@external`, they didn't behaved like other external fields; this was an inconsistency in the use of `@external` in federation 1. Tl;dr, federation 1 ideally would never have required key field to be marked `@external`, but it did.
- for federation 2, we wanted to move toward fixing that inconsistency with `@external`, but we also didn't wanted to break all the old subgraphs. So the compromise we went with is to make type defined with `extend type` a bit special in the sense that `@external` on key field is ignored for those type and those types declaration only. But if you don't use `extend type` (recommended in federation 2), then `@external` means the same thing on all field, including key fields, and it means that the subgraph does not resolve the field (which here, is again not an option due to the `@requires`).

Tl;dr, just remove the `@external` on `Product.id`, or add back the `extend`.

I'll also note that this this what the error message was trying to point to with the `from subgraph "inventory": cannot satisfy @require conditions on field "Product.shippingEstimate" (please ensure that this is not due to key field "id" being accidentally marked @external).` but I admit it is maybe a bit buried and it is on my todo list to try to improve those errors a bit (to simplify them/surface the most likely problem a bit more prominently).

**Comment by pcmanus** [comment]
Note: closing this issue as this is working largely as designed and is mostly a gotcha from moving subgraph from the federation 1 "style" to the federation 2 one. But I'll note that if you really invented here to make `Product.id` to be a "true" `@external` (which again, it was _not_ before when the type was declared with `extend`) then we can discuss further, but I'll need to understand why you want that (I'm unsure why you would want that in that context though).
